### PR TITLE
Avoid reloading _source for every inner hit.

### PR DESCRIPTION
**Pull request by jtibshirani**
Previously if an inner_hits block required _ source, we would reload and parse
the root document's source for every hit. This PR adds a shared SourceLookup to
the inner hits context that allows inner hits to reuse parsed source if it's
already available. This matches our approach for sharing the root document ID.

Relates to #32818.

### CODE REVIEW

**Comment by jtibshirani** [comment]
I ran benchmarks against the `nested` track in rally before and after the change. The service time improves when loading a large number of `inner_hits`. I would expect the change to be more significant with a larger document _source (the documents in the `nested` track are quite small).

```
esrally --pipeline=benchmark-only --track-path=nested --challenge=nested-search-challenge \
    --track-params="ingest_percentage:10"
```

```
BASELINE
| 50th percentile service time | randomized-nested-queries-with-inner-hits_default_big_size | 50.1528 | ms |
| 90th percentile service time | randomized-nested-queries-with-inner-hits_default_big_size | 60.7729 | ms |
| 99th percentile service time | randomized-nested-queries-with-inner-hits_default_big_size | 74.1306 | ms |

CONTENDER
| 50th percentile service time | randomized-nested-queries-with-inner-hits_default_big_size | 43.5691 | ms |
| 90th percentile service time | randomized-nested-queries-with-inner-hits_default_big_size | 50.9585 | ms |
| 99th percentile service time | randomized-nested-queries-with-inner-hits_default_big_size | 62.4088 | ms |
```

File: server/src/main/java/org/elasticsearch/search/fetch/subphase/InnerHitsContext.java [hunk 1]
```diff
@@ -102,22 +105,43 @@ public void setChildInnerHits(Map<String, InnerHitSubContext> childInnerHits) {
             this.childInnerHits = new InnerHitsContext(childInnerHits);
         }
 
-        protected Weight createInnerHitQueryWeight() throws IOException {
-            final boolean needsScores = size() != 0 && (sort() == null || sort().sort.needsScores());
-            return context.searcher().createWeight(context.searcher().rewrite(query()),
+        protected Weight getInnerHitQueryWeight() throws IOException {
```
Comment (jtibshirani) [inline server/src/main/java/org/elasticsearch/search/fetch/subphase/InnerHitsContext.java:108]: This felt a bit hacky. We take a similar strategy in `FetchSourceContext#getFilter`.

Comment (jimczi) [reply server/src/main/java/org/elasticsearch/search/fetch/subphase/InnerHitsContext.java:108]: I think it's still hacky but slightly better.
There should be a way to share a state in `FetchSubPhase#hitExecute` though. 
If we plan to use `hitExecute` more, maybe we should provide a way to  share a state between hits ?
A factory could help for instance:
`````
Function<SearchContext, FetchSubPhase> fetchSubPhase;
`````

**Review by jimczi — APPROVED** [review]
I think it's a great simplification. I left one comment regarding the usage of `hitExecute` in general but that's more an idea for a follow up.

File: server/src/main/java/org/elasticsearch/search/fetch/subphase/InnerHitsContext.java (hunk 1 above)
Comment (jtibshirani) [reply server/src/main/java/org/elasticsearch/search/fetch/subphase/InnerHitsContext.java:108]: This is a good idea, I do think we'll need a mechanism like this as we convert more phases to use `hitExecute`.

**Pull request merged**